XNAControls/XNAScrollPanel.cs

```diff
@@ -0,0 +1,658 @@
+            Y = AllowScroll.Y
+                && (isOverflowingHard.Y || isOverflowingSoft.Y && isOverflowingHard.X),
+        };
+        
+        (HorizontalScrollBar.Visible, VerticalScrollBar.Visible) = scrollbarVisible;
+        CornerPanel.Visible = scrollbarVisible.X && scrollbarVisible.Y;
+
+        #endregion
+        
+        HorizontalScrollBar.ClientRectangle = new()
+        {
+            X = border,
+            Y = Height
+                - HorizontalScrollBar.ScrollHeight 
+                - border,
+            Width = Width
+                - border2x
+                - (scrollbarVisible.Y ? VerticalScrollBar.ScrollWidth : 0),
+            Height = HorizontalScrollBar.ScrollHeight,
+        };
+
+        VerticalScrollBar.ClientRectangle = new()
+        {
+            X = Width
+                - VerticalScrollBar.ScrollWidth 
+                - border,
+            Y = border,
+            Width = VerticalScrollBar.ScrollWidth,
+            Height = Height 
+                - border2x
+                - (scrollbarVisible.X ? HorizontalScrollBar.ScrollHeight : 0),
+        };
+
+        // keep in mind that CurrentViewSize call here relies on correct placement of scrollbars
+        CornerPanel.ClientRectangle = new()
+        {
+            Location = ViewSize,
+            Size = ClientRectangle.Size - ViewSize - new Point(border),
+        };
+        
+        HorizontalScrollBar.DisplayedPixelCount = ViewSize.X;
+        VerticalScrollBar.DisplayedPixelCount = ViewSize.Y;
+        
+        HorizontalScrollBar.Refresh();
+        VerticalScrollBar.Refresh();
+    }
+    
+    #endregion
+
+    #region Scroll methods
+
+    /// <summary>
+    /// Scrolls to the specified rectangle.
+    /// </summary>
+    /// <remarks>
+    /// If the rectangle is bigger than the viewport - scrolls to it's top/left bounds.
+    /// </remarks>
+    /// <param name="rect">The rectangle (in local coordinates) to scroll to.</param>
+    public virtual void ScrollTo(Rectangle rect)
+    {
+        // Math.Min call inside the calculation is responsible for handling controls bigger
+        // than the viewport (basically makes the viewport snap to top left corner, or top/left
+        // sides separately if it overflows on only one side).
+        
+        // As an alternative implementation you might want to have the viewport scroll to
+        // the closest part of the control. To do that simply flip min and max values or flip
+        // places of CurrentViewRectangle and rect in the calculation.
+        
+        CurrentViewPosition = new()
+        {
+            X = Clamp(value: CurrentViewRectangle.X,
+                min: Math.Min(rect.X + rect.Width - CurrentViewRectangle.Width, rect.X),
+                max: rect.X),
+            Y = Clamp(value: CurrentViewRectangle.Y,
+                min: Math.Min(rect.Y + rect.Height - CurrentViewRectangle.Height, rect.Y),
+                max: rect.Y),
+        };
```

Comment:
Another implementation to consider here is that when the rectangle to scroll to is smaller than the viewport, the rectangle could be centered within the viewport.

Reply:
In my experience I haven't seen such behavior used in generic UI controls. I was mainly thinking of a common use case of switching focus through children using `[Tab]`, IIRC in almost all Windows apps it does minimal amount of scrolling to show the control fully.

But I can see this making sense for some specific controls, like a map view, which would center on the point you want to focus to. I think such controls should implement that in an override maybe? Doesn't hurt to add this to the comments above though, if that's what you meant.